torch/profiler/_utils.py

```diff
@@ -18,32 +42,214 @@ def __eq__(self, other):
+            self_time = curr_event.duration_time_ns
+            for child_event in curr_event.children:
+                self_time -= child_event.duration_time_ns
+                stack.append(child_event)
+
+            assert EventKey(
+                curr_event
+            ) not in self.metrics, f"Duplicate id: {curr_event.id}, {curr_event.name()}"
+            self.metrics[EventKey(curr_event)] = EventMetrics(
+                self_time_ns=self_time)
+            self.metrics[EventKey(
+                curr_event)].duration_time_ns = curr_event.duration_time_ns
+
+    def compute_idle_time_and_qd(self):
+        '''
+        Computes event's idle time. Idle time is defined as the time when the CUDA kernel queue depth is 0.
+        It also return a Time series of the queue depth data.
+        qd = cuda kernel queue depth
+        '''
+        assert (self.profile.kineto_results is not None)
+        event_list = self.profile.kineto_results.events()
+
+        def is_cuda_launch_kernel(e):
+            # TODO: find a better way to identify cudaLaunchKernel
+            return e.name() == "cudaLaunchKernel"
 
+        def is_cuda_kernel(e):
+            # TODO: find a better way to identify CUDA Kernel
+            return e.device_type() == DeviceType.CUDA and "mem" not in e.name(
+            ).lower()
 
-def compute_self_time(prof: profile, metrics: Dict[EventKey, EventMetrics]):
-    '''
-    Computes event's self time(total time - time in child ops).
+        # Record All the idle intervals
+        curr_qd = 0
+        idle_interval = []
+        qd_list = []
+        idle_start = 0
+        cuda_kernel_events = [
+            event for event in event_list
+            if is_cuda_launch_kernel(event) or is_cuda_kernel(event)
+        ]
+        cuda_kernel_events.sort(key=lambda e: e.start_us())
+
+        for curr_event, next_event in zip(cuda_kernel_events,
+                                          cuda_kernel_events[1:]):
+            if (is_cuda_launch_kernel(curr_event)):
+                if (curr_qd == 0):
+                    if idle_start is not None:
+                        idle_interval.append(
+                            Interval(idle_start * 1000,
+                                     curr_event.start_us() * 1000))
+                curr_qd += 1
+            if (is_cuda_kernel(curr_event)):
+                curr_qd -= 1
+                if (curr_qd == 0):
+                    idle_start = curr_event.start_us()
+
+            qd_list.append(
+                Interval(curr_event.start_us() * 1000,
+                         (next_event.start_us() + next_event.duration_us()) *
+                         1000, curr_qd))
+
+        event_list = [e.event for e in self.metrics.keys()]
+        for event in event_list:
+            self.metrics[EventKey(event)].idle_time_ns = EventKey(
+                event).intervals_overlap(idle_interval)
+        return qd_list
+
+    def rank_events(self, length):
```

Comment:
This is really a second PR, so I'm going to defer and recommend that you split this out.